### DISCUSSION TITLE
Is there any way to call the backend without freezing the web window?

### DISCUSSION THREAD

**Discussion by Xoffio** [tauri-apps/tauri#4191] [Q&A] [answered]
Hi,

First of all, Thanks you so much for the Tauri team. This code really is a game changer ! 

I am using Tauri to load some images from rust. My front end code looks something like this:

```js
function startCaching(){

        for (let nFrame=0; nFrame<9; nFrame++){

            imageNumber = 1 + nFrame;
            invoke('get_image_raw_data', { imageNumber, canvasW, canvasH }).then((data_from_rust) => {
                rawImagesFrames.push(data_from_rust);
            })
        }
    }
```

And my backend code looks something like this:

```rust
#[tauri::command]
fn get_image_raw_data(image_number: i32, canvas_w: f32, canvas_h: f32) -> Result<Vec<u8>, String> {
  use image::GenericImageView;
  
  ...
  
  // Return raw image
  Ok(img_raw_data)
```

The problem is that when I call `get_image_raw_data` the main window freezes until I get all my data. Is this a bug or am I doing something wrong ?

P.S:

I found this [issue](https://github.com/tauri-apps/tauri/issues/3421) that might be the same problem I am having. If this is actually a bug, is there any workaround while I white for the tauri team to solve it?

**Comment by FabianLars** [comment, marked as answer]
Make the command async, either by changing it to `async fn` or `#[tauri::command(async)]`. This way it doesn't run on the main thread.

**Reply by Xoffio** [reply]
@FabianLars Thank you for the guidance ! I really appreciated. Once I make the function in rust `async` does the function `invoke` becomes a promise? In my code I have a `for` loop getting multiple frames from rust. How can I wait until I get all the frames? sorry for ask it just that I got a bit confuse

**Reply by FabianLars** [reply]
Every invoke call is always a promise, no matter if the rust command is sync or async, because the communication with the backend itself is async. That's why you have to use `.then` in your example

**Reply by Xoffio** [reply]
Yeah, Got it! JS is not my strength. 

One thing. After I changed the function to `async` the window was more responsive at the beginning. One thing I noticed is that the window still getting unresponsive when `invoke` gets the array of data from rust.... :( This lag is right when `invoke` gets the data. Let's say I am getting 3 images from rust. The window will get laggy 3 times right when getting the data... Do you think this is Tauri? or I am missing something else? I started to look for a solution on the internet and I found [this](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) do you think with `workers` I can call `invoke` in a different thread in the JS side so to avoid the windows becoming unresponsive?

Once again. Thank you so much!!

**Reply by FabianLars** [reply]
That's the IPC being a bottleneck here, it's even worse for binary data. For images in particular, the current workaround is to use the asset protocol via the convertFileSrc function (this needs the images on the filesystem tho).

> I found [this](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) do you think with workers I can call invoke in a different thread in the JS side so to avoid the windows becoming unresponsive?

You don't have access to tauri apis in worker, but i doubt that this would help anyway since the whole webview is freezed by the ipc channel.

P.S. We will rewrite the IPC for v2, but many solutions are still blocked by platform limitations (mainly on linux)

**Reply by Xoffio** [reply]
Hey @FabianLars ! Thanks for taking the time !!! 

>  For images in particular, the current workaround is to use the asset protocol via the convertFileSrc function (this needs the images on the filesystem tho).

I was reading [convertFileSrc](https://tauri.studio/v1/api/js/modules/tauri/). (Correct me if I am wrong) So what this function does is load the image from JS?. The problem is that I am having rust decoding and doing modifications to the image, basically doing all the hard work in the back end (which is working excellent) So I don't think JS will help me with this... I guess I can try to send the images in pieces to see if that helps... Maybe I can find the right amount of data I can send from rust to js with out bottlenecking the IPC?

> You don't have access to tauri apis in worker, but i doubt that this would help anyway since the whole webview is freezed by the ipc channel.

If importing taui's API from the webworker were possible. Wouldn't the that thread the one getting bottleneck instated of the main one?

> P.S. We will rewrite the IPC for v2, but many solutions are still blocked by platform limitations (mainly on linux)

OHH Man Happy to heard that. You guys are doing an amazing work!!! Thanks you all!

**Reply by FabianLars** [reply]
> I was reading [convertFileSrc](https://tauri.studio/v1/api/js/modules/tauri/). (Correct me if I am wrong)...

The function itself just gives you a modified file path which tauri can interpret to load files into the webview. It's really just a different "channel" to load data into the webview. Instead of the IPC it uses a custom protocol (like the html,js,css files themselves), the custom protocol can handle binary data just fine (mainly because it doesn't need to serialize it to a string).

> If importing taui's API from the webworker were possible. Wouldn't the that thread the one getting bottleneck instated of the main one?

Best case, yes. But from what i see it would have the same problem because the IPC currently blocks the whole webview and that should include workers, but not sure about that. Another problem is that you don't have access to the `window` object which is needed by the IPC so we'd have to route the request through the main process anyway...

> OHH Man Happy to heard that. You guys are doing an amazing work!!! Thanks you all!

❤️

**Reply by Xoffio** [reply]
Hi @FabianLars. I was writing the code to send the image in chucks but then I realize that this will also not work because I would have to load the image multiple times... 

Anyway, doing some research I found [this issue](https://github.com/tauri-apps/tauri/issues/1817) where someone have similar problem. @lucasfernog posted a [link](https://github.com/tauri-apps/tauri-invoke-http) to open a port and communicate with the front end. I think this can solve my problem while I wait for v2 to come out. One last question, since you know the core code a lot lot more than me. Do you think using `tauri-invoke-http` would be better than something like rocket to send the images array through http? I am going to have similar problem using `tauri-invoke-http`?

Once again -> Thank you so much !!!

**Reply by FabianLars** [reply]
tauri-invoke-http is faster than the main ipc, but not as fast as it could be since the payload still gets serialized to a string first.
So yeah, a completely separate system would be faster than invoke-http, yes, but using rocket for this sounds slightly overkill/too heavy to me :P

**Reply by Xoffio** [reply]
Great to heard that! I haven't used rocket. I was just searching on the internet for something that can help me solve this problem and that's how I found it. If you know something less overkill I would love to know. Thanks!

**Reply by FabianLars** [reply]
Well, my current go-to library (both in usage and recommendation) is axum, and before that came a long i liked warp too.

Lately I'm really going crazy and use the std TcpListener on way too many occasions, so if you like the pain, that could be an alternative too.

But yeah rocket may be a good choice if you're used to server frameworks in other high-level languages, probably a good starting point and it should be relatively easy to swap it out if needed.

**Reply by Xoffio** [reply]
I'll just post this code here in case someone has the same problem as me.

```rust
use axum::{
  routing::{get},
  Router,
};
use std::net::SocketAddr;

#[tokio::main]
async fn main() {

  tokio::spawn(async {
    // build our application with a route
    let app = Router::new()
    // `GET /` goes to `root`
    .route("/", get(root));
  
    // run our app with hyper
    // `axum::Server` is a re-export of `hyper::Server`
    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    println!("listening on {}", addr);
    //tracing::debug!("listening on {}", addr);
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
  });

  tauri::Builder::default()
    .invoke_handler(tauri::generate_handler![get_image_raw_data])
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}

// basic handler that responds with a static string
async fn root() -> &'static str {
  "This should return my raw Image"
}
```

@FabianLars I decided to go for `Axum` as your recommendation. For now this seems to be working. Tomorrow I'll try with an image an see the performance. @FabianLars Do you see anything wrong doing it this way? I had to make the main function `async` and call `axum` with a thread. I saw in this [discussion](https://github.com/tauri-apps/tauri/discussions/2942) you did it differently but I coun't replicated here.. After this Ill leave you alone. Thank you so much for all the help !!

**Reply by FabianLars** [reply]
Yep that's totally fine, you just need to give tauri a handle to the tokio runtime so that it won't try to spawn it's own (would result in a crash).
https://docs.rs/tauri/1.0.0-rc.13/tauri/async_runtime/fn.set.html
